trin-core/src/portalnet/types/messages.rs

```diff
@@ -619,272 +570,205 @@ impl FromStr for HexData {
+        assert_eq!(encoded, expected_encoded);
 
-    #[test]
-    #[should_panic]
-    fn invalid_protocol_id() {
-        let hex_string = "500F";
-        ProtocolId::from_str(hex_string).unwrap();
+        let decoded = Message::try_from(hex::decode(expected_encoded).unwrap()).unwrap();
+        assert_eq!(decoded, nodes);
     }
 
-    // test vectors sourced from
-    // https://github.com/ethereum/portal-network-specs/blob/master/portal-wire-test-vectors.md
     #[test]
-    fn test_vector_ping() {
-        let enr_seq = 1u64;
-        let data_radius = U256::MAX - U256::from(1);
-        let custom_payload = CustomPayload::from(data_radius.as_ssz_bytes());
-        let expected = "0001000000000000000c000000feffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff";
-        let message = Message::Request(Request::Ping(Ping {
-            enr_seq,
-            custom_payload,
-        }));
-        assert_eq!(hex::encode(message.to_bytes()), expected);
-        let decoded = Message::from_bytes(message.to_bytes().as_slice()).unwrap();
-        assert_eq!(decoded, message);
-    }
+    fn message_encoding_find_content() {
+        let content_key = hex::decode("706f7274616c").unwrap();
+        let find_content = FindContent { content_key };
+        let find_content = Message::FindContent(find_content);
 
-    #[test]
-    fn test_vector_pong() {
-        let enr_seq = 1;
-        let data_radius: U256 = U256::max_value() / 2;
-        let custom_payload = CustomPayload::from(data_radius.as_ssz_bytes());
-        let expected = "0101000000000000000c000000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff7f";
-        let message = Message::Response(Response::Pong(Pong {
-            enr_seq,
-            custom_payload,
-        }));
-        assert_eq!(hex::encode(message.to_bytes()), expected);
-        let decoded = Message::from_bytes(message.to_bytes().as_slice()).unwrap();
-        assert_eq!(decoded, message);
-    }
+        let encoded: Vec<u8> = find_content.clone().into();
+        let encoded = hex::encode(encoded);
+        let expected_encoded = "0404000000706f7274616c";
+        assert_eq!(encoded, expected_encoded);
 
-    #[test]
-    fn test_vector_find_nodes() {
-        let distances = vec![256, 255];
-        let expected = "02040000000001ff00";
-        let message = Message::Request(Request::FindNodes(FindNodes { distances }));
-        assert_eq!(hex::encode(message.to_bytes()), expected);
-        let decoded = Message::from_bytes(message.to_bytes().as_slice()).unwrap();
-        assert_eq!(decoded, message);
+        let decoded = Message::try_from(hex::decode(expected_encoded).unwrap()).unwrap();
```

Comment:
*nitpick* on one hand you just validated that `expected_encoded == encoded` on the other hand it makes more sense to me to use the tested variable for the round trip test (aka `hex::decode(encoded)`)

Reply:
I see what you mean. My thinking was that we should decode from "ground truth" encoding, but like you said we just validated its equality to `encoded`. And I do think it reads more nicely the way you've proposed.